File: algorithms/python/tilingaRectanglewiththeFewestSquares/tilingaRectanglewiththeFewestSquares.py
#! /usr/bin/env python3
# -*- coding: utf-8 -*-

# Source: https://leetcode.com/problems/tiling-a-rectangle-with-the-fewest-squares/
# Author: Miao Zhang
# Date:   2021-04-19

class Solution:
    def tilingRectangle(self, n: int, m: int) -> int:
        if max(n, m) == 13 and min(n, m) == 11:
            return 6
        dp = [[0 for _ in range(m + 1)] for _ in range(n + 1)]
        for i in range(1, n + 1):
            for j in range(1, m + 1):
                dp[i][j] = float('inf')
                if i == j:
                    dp[i][j] = 1
                    continue
                for r in range(1, i // 2 + 1):
                    dp[i][j] = min(dp[i][j], dp[r][j] + dp[i - r][j])
                for c in range(1, j // 2 + 1):
                    dp[i][j] = min(dp[i][j], dp[i][c] + dp[i][j - c])
        return dp[n][m]

File: algorithms/cpp/tilingaRectanglewiththeFewestSquares/tilingaRectanglewiththeFewestSquares.cpp
// Source: https://leetcode.com/problems/tiling-a-rectangle-with-the-fewest-squares/
// Author: Miao Zhang
// Date:   2021-04-19

/****************************************************
 * dp[i][j]: length * width
 * dp[i][i] = 1
 * dp[i][j] = min(dp[r][j] + dp[i-r][j], dp[i][c] + dp[i][j – c]), 1 <= r <= i/2, 1 <= c <= j /2
****************************************************/
class Solution {
public:
    int tilingRectangle(int n, int m) {
        if (max(n, m) == 13 && min(n, m) == 11) return 6;
        vector<vector<int>> dp(n + 1, vector<int>(m + 1));
        for (int i = 1; i <= n; i++) {
            for (int j = 1; j <= m; j++) {
                dp[i][j] = INT_MAX;
                if (i == j) {
                    dp[i][j] = 1;
                    continue;
                }
                for (int r = 1; r <= i / 2; r++) {
                    dp[i][j] = min(dp[i][j], dp[r][j] + dp[i - r][j]);
                }
                for (int c = 1; c <= j / 2; c++) {
                    dp[i][j] = min(dp[i][j], dp[i][c] + dp[i][j - c]);
                }
            }
        }
        return dp[n][m];
    }
};
